app/src/main/java/com/google/android/sambadocumentsprovider/encryption/EncryptionManager.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.google.android.sambadocumentsprovider.encryption;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.security.keystore.KeyGenParameterSpec;
+import android.security.keystore.KeyProperties;
+import android.util.Base64;
+import android.util.Log;
+
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.NoSuchProviderException;
+import java.security.cert.CertificateException;
+import java.util.Random;
+
+import javax.crypto.Cipher;
+import javax.crypto.KeyGenerator;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.GCMParameterSpec;
+
+public class EncryptionManager {
+  private static final String TAG = "EncryptionManager";
+
+  private static final String ENCRYPTION_MANAGER_PREF_KEY = "encryptionManager";
+  private static final String KEYSTORE_PROVIDER = "AndroidKeyStore";
+  private static final String KEY_ALIAS = "SambaEncryptionKey";
+  private static final String AES_CIPHER = KeyProperties.KEY_ALGORITHM_AES + "/" +
+          KeyProperties.BLOCK_MODE_GCM + "/" +
+          KeyProperties.ENCRYPTION_PADDING_NONE;
+  private static final int GCM_TAG_LENGTH = 128;
+  private static final int IV_LENGTH = 12;
+  private static final String DEFAULT_CHARSET = "UTF-8";
+
+  private static final Random mRandom = new Random();
+
+  private final SharedPreferences mPref;
+  private EncryptionKey mKey;
+
+  private KeyStore mStore;
```

Comment:
Oh no... I just saw that. First of all, try not to catch generic Exception unless you have strong reason. If an exception is thrown, this class will generate tons of NPE so it should convert exceptions that shouldn't happen at all into RuntimeException, and keep other exceptions thrown to let outside handle it.